crates/ruff_linter/src/rules/pylint/rules/unspecified_encoding.rs

```diff
@@ -0,0 +1,118 @@
+///
+/// ## Why is this bad?
+/// Using `open` in text mode without an explicit encoding specified can lead to
+/// unportable code that leads to different behaviour on different systems.
+///
+/// Instead, consider using the `encoding` parameter to explicitly enforce a specific encoding.
+///
+/// ## Example
+/// ```python
+/// open("file.txt")
+/// ```
+///
+/// Use instead:
+/// ```python
+/// open("file.txt", encoding="utf-8")
+/// ```
+///
+/// ## References
+/// - [Python documentation: `open`](https://docs.python.org/3/library/functions.html#open)
+#[violation]
+pub struct UnspecifiedEncoding {
+    function_name: String,
+}
+
+impl Violation for UnspecifiedEncoding {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!(
+            "`{}` {}without explicit `encoding` argument",
+            self.function_name,
+            if self.function_name == "open" {
+                "in text mode "
+            } else {
+                ""
+            }
+        )
+    }
+}
+
+fn is_binary_mode(expr: &ast::Expr) -> Option<bool> {
+    Some(expr.as_constant_expr()?.value.as_str()?.value.contains('b'))
+}
+
+fn is_violation(call: &ast::ExprCall, path: &[&str]) -> bool {
+    // this checks if we have something like *args which might contain the encoding argument
+    if call
+        .arguments
+        .args
+        .iter()
+        .any(ruff_python_ast::Expr::is_starred_expr)
+    {
+        return false;
+    }
+    // this checks if we have something like **kwargs which might contain the encoding argument
+    if call.arguments.keywords.iter().any(|a| a.arg.is_none()) {
+        return false;
+    }
+    match path {
+        ["" | "codecs", "open"] => {
+            if let Some(mode_arg) = call.arguments.find_argument("mode", 1) {
+                if is_binary_mode(mode_arg).unwrap_or(true) {
+                    // binary mode or unknown mode is no violation
+                    return false;
+                }
+            }
+            // else mode not specified, defaults to text mode
+            call.arguments.find_argument("encoding", 3).is_none()
+        }
+        ["io", "TextIOWrapper"] => call.arguments.find_argument("encoding", 1).is_none(),
+        ["tempfile", "TemporaryFile" | "NamedTemporaryFile" | "SpooledTemporaryFile"] => {
+            let mode_pos = usize::from(path[1] == "SpooledTemporaryFile");
+            if let Some(mode_arg) = call.arguments.find_argument("mode", mode_pos) {
+                if is_binary_mode(mode_arg).unwrap_or(true) {
+                    // binary mode or unknown mode is no violation
+                    return false;
+                }
+            } else {
+                // defaults to binary mode
+                return false;
+            }
+            call.arguments
+                .find_argument("encoding", mode_pos + 2)
+                .is_none()
+        }
+        _ => false,
+    }
```

Comment:
This implementation overall looks really nice, but where this list come from? E.g., it looks like Pylint flags some of the `pathlib` methods -- should we flag those here too?

Reply:
Ahh, I guess we can't include the `pathlib` cases since those are instance methods.

Reply:
Exactly, we tried `pathlib` as well but it didn't work.
We also tried to detect something like this:
```python
foo = open
foo("test.txt")
```
But we didn't manage to get this one right, so we omitted it. =(